test/unit/puppetlabs/puppetserver/lockable_pool_test.clj

```diff
@@ -167,54 +207,314 @@
+          (str "did not get back the same instance from the third borrow"
+               "attempt as was returned from the first borrow attempt")))))
+
+(deftest pool-can-borrow-after-borrow-timed-out-during-lock
+  (testing (str "can do a timed borrow from pool after previous borrow "
+                "timed out while a write lock was held")
+    (let [pool (create-populated-pool 1)
+          borrow-1 (.borrowItem pool)
+          borrow-thread-started-2? (promise)
+          borrow-thread-started-3? (promise)
+          borrow-thread-2 (future
+                           (deliver borrow-thread-started-2? true)
+                           (.borrowItemWithTimeout
+                            pool
+                            50
+                            TimeUnit/MILLISECONDS))
+          borrow-thread-3 (future
+                           (deliver borrow-thread-started-3? true)
+                           (.borrowItemWithTimeout pool
+                                                   1
+                                                   TimeUnit/SECONDS))
+          _ @borrow-thread-started-2?
+          _ @borrow-thread-started-3?
+          unlock? (promise)
+          lock-thread-started? (promise)
+          lock-thread-locked? (promise)
+          lock-thread (future
+                       (deliver lock-thread-started? true)
+                       (.lock pool)
+                       (deliver lock-thread-locked? true)
+                       @unlock?
+                       (.unlock pool))]
+      @lock-thread-started?
+      (.releaseItem pool borrow-1)
+      @lock-thread-locked?
+      (is (true? true))
+      (is (nil? @borrow-thread-2)
+          "second borrow attempt did not time out")
+      (deliver unlock? true)
+      @lock-thread
+      (is (true? true))
+      (is (identical? borrow-1 @borrow-thread-3)
+          (str "did not get back the same instance from the third borrow"
+               "attempt as was returned from the first borrow attempt")))))
+
+(deftest pool-insert-pill-test
+  (testing "inserted pill is next item borrowed"
+    (let [pool (create-populated-pool 1)
+          pill (str "i'm a pill")]
+      (.insertPill pool pill)
+      (is (identical? (.borrowItem pool) pill)))))
+
+(deftest pool-clear-test
+  (testing (str "pool clear removes all elements from queue and only matching"
+                "registered elements")
+    (let [pool (create-populated-pool 3)
+          instance (.borrowItem pool)]
+      (is (= 2 (.size pool)))
+      (is (= 3 (-> pool
+                   (.getRegisteredElements)
+                   (.size))))
+      (.clear pool)
+      (is (= 0 (.size pool)))
+      (let [registered-elements (.getRegisteredElements pool)]
+        (is (= 1 (.size registered-elements)))
+        (is (identical? instance (-> registered-elements
+                                     (.iterator)
+                                     iterator-seq
+                                     first)))))))
```

Comment:
@KevinCorcoran and I also talked about how `clear` should work here - https://github.com/puppetlabs/puppet-server/pull/813#discussion_r48579770.  I made the argument that it would be desirable for `clear` to work this way because it would seem wrong for a borrower to return an instance to the pool after the `clear` call was made and end up in the situation where we have an instance in the `liveQueue` that was not also in `registeredElements`.

We don't actually use `clear` in any production code anyway, AFAICT.  I doubt we ever will either since we need to do some custom cleanup processing on any elements that we want to dump from the pool, which involves borrowing the individual instances.

Does this make sense?